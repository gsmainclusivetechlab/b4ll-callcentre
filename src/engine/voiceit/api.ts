import VoiceIt2, { VoiceIt } from 'voiceit2-nodejs';

export const REQUIRED_ENROLMENTS = 3;

const myVoiceIt = new VoiceIt2(
    process.env.VOICEIT_API_KEY || '',
    process.env.VOICEIT_API_TOKEN || ''
);

export async function createUser(): Promise<VoiceIt.Response<VoiceIt.User>> {
    return new Promise<VoiceIt.Response<VoiceIt.User>>((resolve) => {
        myVoiceIt.createUser((callback) => {
            resolve(callback);
        });
    });
}
interface ArrayEnrolment {
    createdAt: string;
    contentLanguage: string;
    voiceEnrollmentId: number;
    text: string;
}

export async function getPhrases(
    language: VoiceIt.ContentLanguage
): Promise<
    VoiceIt.Response<{
        count: number;
        phrases: VoiceIt.Phrase[];
    }>
> {
    return new Promise<
        VoiceIt.Response<{
            count: number;
            phrases: VoiceIt.Phrase[];
        }>
    >((resolve) => {
        myVoiceIt.getPhrases(
            {
                contentLanguage: language,
            },
            (callback) => {
                resolve(callback);
            }
        );
    });
}

export async function getEnrolledPhrases(
    userId: string,
    language: string
): Promise<Record<string, number>> {
    return new Promise<Record<string, number>>((resolve) => {
        myVoiceIt.getAllVoiceEnrollments(
            {
                userId: userId,
            },
            (callback) => {
                const phrases = ((callback.voiceEnrollments as unknown) as ArrayEnrolment[])
                    .map(
                        ([
                            id,
                            encodedPhrase,
                            contentLanguage,
                            createdAt,
                        ]): VoiceIt.VoiceEnrolment => ({
                            voiceEnrollmentId: id,
                            contentLanguage,
                            createdAt,
                            text: Buffer.from(encodedPhrase, 'base64').toString(
                                'utf8'
                            ),
                        })
                    )
                    .filter((x) => x.contentLanguage === language)
                    .map((x) => x.text);
                resolve(
                    phrases.reduce(
                        (acc, curr) => ({
                            ...acc,
                            [curr]: (acc[curr] || 0) + 1,
                        }),
                        {} as Record<string, number>
                    )
                );
            }
        );
    });
}

export async function enrolUser(
    userId: string,
    language: VoiceIt.ContentLanguage,
    enrolment: {
        phrase: string;
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<VoiceIt.VoiceEnrolment>> {
    return new Promise<VoiceIt.Response<VoiceIt.VoiceEnrolment>>((resolve) => {
        myVoiceIt.createVoiceEnrollmentByUrl(
            {
                userId: userId,
                contentLanguage: language,
                phrase: enrolment.phrase,
                audioFileURL: enrolment.recordingUrl,
            },
            (jsonResponse) => {
                resolve(jsonResponse);
            }
        );
    });
}

export async function verifyUser(
    userId: string,
    language: VoiceIt.ContentLanguage,
    enrolment: {
        phrase: string;
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<{ confidence: number }>> {
    return new Promise<VoiceIt.Response<{ confidence: number }>>((resolve) => {
        myVoiceIt.voiceVerificationByUrl(
            {
                userId,
                phrase: enrolment.phrase,
                audioFileURL: enrolment.recordingUrl,
                contentLanguage: language,
            },
            (callback) => {
                resolve(callback);
            }
        );
    });
}
