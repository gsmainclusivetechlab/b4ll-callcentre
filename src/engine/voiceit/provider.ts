import {
    BiometricType,
    EnrolmentRequest,
    BiometricsProvider,
} from '../BiometricsProvider';
import {
    createUser,
    getEnrolledPhrases,
    REQUIRED_ENROLMENTS,
    getPhrases,
    enrolUser,
    verifyUser,
} from './api';

interface Params {
    userId: string;
    language: string;
}

declare const voiceItData: unique symbol;

export type VoiceItEnrolmentData = {
    phrase: string;
    recordingsRequired: number;
    userId: string;
} & {
    // prevent manual creation of EnrolmentData, which is not the intended use
    [voiceItData]: true;
};

export type VoiceItVerificationData = {
    phrase: string;
} & {
    // prevent manual creation of VerificationData, which is not the intended use
    [voiceItData]: true;
};

export type VoiceItProvider = BiometricsProvider<
    BiometricType.VOICE,
    VoiceItEnrolmentData,
    Omit<Params, 'userId'> & { userId?: string },
    Params,
    VoiceItVerificationData,
    Params,
    Params
>;

export const provider: VoiceItProvider = {
    getEnrolmentRequest: async ({ userId, language }) => {
        if (!userId) {
            const voiceItUser = await createUser();
            userId = voiceItUser.userId;
        }
        const phraseCounts = await getEnrolledPhrases(userId, language);

        // find whichever phrase has most enrolments (if any)
        const unfinishedPhrase = Object.keys(phraseCounts).reduce(
            (curr, phrase) =>
                curr.count < phraseCounts[phrase]
                    ? { count: phraseCounts[phrase], phrase }
                    : curr,
            { count: 0, phrase: undefined as string | undefined }
        );

        let phrase: string, recordingsRequired: number;
        if (unfinishedPhrase.phrase && unfinishedPhrase.count < 3) {
            phrase = unfinishedPhrase.phrase;
            recordingsRequired = Math.max(
                1,
                REQUIRED_ENROLMENTS - unfinishedPhrase.count
            );
        } else {
            const phrases = await getPhrases(language);
            const availablePhrases = phrases.map((x) => x.text.toLowerCase());
            phrase =
                availablePhrases[
                    Math.floor(Math.random() * availablePhrases.length)
                ];
            recordingsRequired = REQUIRED_ENROLMENTS;
        }
        if (!phrase) throw new Error('No phrases available for enrolment');
        return {
            biometricType: BiometricType.VOICE,
            phrase: phrase,
            request: {
                phrase,
                userId,
                recordingsRequired,
            } as VoiceItEnrolmentData,
        };
    },

    handleEnrolmentResponse: async (response, { userId, language }) => {
        if (response.biometricType !== BiometricType.VOICE) {
            throw new Error('Unexpected biometric type');
        }
        const {
            recordingUrl,
            request: { phrase },
        } = response;
        const dupeRequest: EnrolmentRequest<
            VoiceItEnrolmentData
        >[BiometricType.VOICE] = {
            biometricType: BiometricType.VOICE,
            phrase: response.request.phrase,
            request: response.request,
        };
        try {
            const existingPhrasePromise = getEnrolledPhrases(userId, language);
            const enrolment = await enrolUser(userId, language, {
                phrase,
                recordingUrl,
            });
            if (enrolment.responseCode === 'SUCC') {
                const existingPhrases = await existingPhrasePromise;
                const phraseCount =
                    (existingPhrases[phrase.toLowerCase()] ?? 0) + 1;

                if (phraseCount < REQUIRED_ENROLMENTS) {
                    dupeRequest.request.recordingsRequired -= 1;

                    return {
                        success: true,
                        complete: false,
                        // repeat the request
                        next: dupeRequest,
                    };
                }
                return { success: true, complete: true };
            }
        } catch (e) {
            // ignore errors and simply retry the request
        }
        return {
            success: false,
            complete: false,
            next: dupeRequest,
        };
    },

    getVerificationRequest: async ({ userId, language }) => {
        const enrolments = await getEnrolledPhrases(userId, language);
        const phrases = Object.keys(enrolments).filter(
            (x) => enrolments[x] >= REQUIRED_ENROLMENTS
        );
        const phrase = phrases[Math.floor(Math.random() * phrases.length)];
        if (!phrase) {
            throw new Error('No phrases available for verification');
        }

        return {
            phrase,
            biometricType: BiometricType.VOICE,
            request: { phrase } as VoiceItVerificationData,
        };
    },

    handleVerificationResponse: async (response, { userId, language }) => {
        if (response.biometricType !== BiometricType.VOICE) {
            throw new Error('Unexpected biometric type');
        }
        const verification = await verifyUser(userId, language, {
            recordingUrl: response.recordingUrl,
            phrase: response.request.phrase,
        });
        const verified = verification.responseCode === 'SUCC';
        return {
            success: verified,
            complete: verified,
            confidence: verification.confidence,
        };
    },
};
