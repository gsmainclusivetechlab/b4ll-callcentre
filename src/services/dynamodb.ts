import AWS from 'aws-sdk';

export interface RecordType {
    count?: number;
    recordingUrl?: string;
}

const getClient = () =>
    new AWS.DynamoDB.DocumentClient({
        endpoint: process.env.DYNAMO_HOST,
        region: process.env.AWS_REGION || 'eu-west-2',
    });

export async function getItem(
    id: string
): Promise<AWS.DynamoDB.DocumentClient.AttributeMap> {
    return getClient()
        .get({
            TableName: process.env.TABLE_NAME || '',
            Key: {
                id,
            },
        })
        .promise()
        .then(({ Item }): RecordType => Item || {});
}

export async function putItem(
    id: string,
    data: RecordType
): Promise<AWS.DynamoDB.DocumentClient.PutItemOutput> {
    return getClient()
        .put({
            TableName: process.env.TABLE_NAME || '',
            Item: {
                id,
                ...data,
            },
        })
        .promise();
}
