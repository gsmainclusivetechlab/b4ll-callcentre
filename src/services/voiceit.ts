import { VoiceIt } from 'voiceit2-nodejs';
import axios from 'axios';
import { SupportedLanguage } from './strings';

const auth = {
    username: process.env.VOICEIT_API_KEY || '',
    password: process.env.VOICEIT_API_TOKEN || '',
};
const api = axios.create({
    baseURL: 'https://api.voiceit.io/',
    auth,
});

export const AVAILABLE_VOICEPRINTS = {
    'en-GB': {
        Tomorrow: 'Never forget tomorrow is a new day',
        Zoo: 'Zoos are filled with small and large animals',
        Telecom: 'please log in to my telecom services account',
    },
    'en-DEV': {
        // These will get mocked anyway so we can use whatever
        Tomorrow: 'Never forget tomorrow is a new day',
        Zoo: 'Zoos are filled with small and large animals',
        Telecom: 'please log in to my telecom services account',
    },
    'fr-FR': {
        Tomorrow: "N'oubliez jamais que demain est un nouveau jour",
        //  TODO: add voiceprints after they are set up in VoiceIt
    },
};

export type LegalVoicePrint<
    L extends SupportedLanguage
> = keyof typeof AVAILABLE_VOICEPRINTS[L];

export function isLegalPhraseKey<L extends SupportedLanguage>(
    language: L,
    phraseKey: unknown
): phraseKey is LegalVoicePrint<L> {
    if (typeof phraseKey !== 'string') return false;
    return phraseKey in AVAILABLE_VOICEPRINTS[language];
}

export async function createUser(): Promise<VoiceIt.Response<VoiceIt.User>> {
    const result = await api.post('/users');
    return result.data;
}

export async function enrolUser<L extends SupportedLanguage>(
    userId: string,
    language: L,
    enrolment: {
        phrase: keyof typeof AVAILABLE_VOICEPRINTS[L];
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<VoiceIt.VoiceEnrolment>> {
    const matches = AVAILABLE_VOICEPRINTS[language] as Record<string, string>;
    const phrase = matches[enrolment.phrase as string];
    const result = await api.post('/enrollments/voice/byUrl', {
        userId,
        phrase,
        fileUrl: enrolment.recordingUrl,
        contentLanguage: 'en-US',
        // TODO: Restore language when we've got a full account
        // contentLanguage: language,
    });
    return result.data;
}

export async function verifyUser<L extends SupportedLanguage>(
    userId: string,
    language: L,
    enrolment: {
        phrase: keyof typeof AVAILABLE_VOICEPRINTS[L];
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<{ confidence: number }>> {
    const matches = AVAILABLE_VOICEPRINTS[language] as Record<string, string>;
    const phrase = matches[enrolment.phrase as string];
    const result = await api.post('/verification/voice/byUrl', {
        userId,
        phrase,
        fileUrl: enrolment.recordingUrl,
        contentLanguage: 'en-US',
        // TODO: Restore language when we've got a full account
        // contentLanguage: language,
    });
    return result.data;
}
