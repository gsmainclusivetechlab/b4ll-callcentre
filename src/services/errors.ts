import { APIGatewayProxyHandler, APIGatewayProxyEvent } from 'aws-lambda';
import {
    getVoiceParams,
    __,
    isSupportedLanguage,
    SupportedLanguage,
} from './strings';
import { twiml } from 'twilio';
import VoiceResponse from 'twilio/lib/twiml/VoiceResponse';
import { getItem, RecordType } from './dynamodb';
import qs from 'querystring';

export interface ParsedRequest {
    language: SupportedLanguage;
    user: RecordType;
    event: APIGatewayProxyEvent;
}

async function parseRequest(
    event: APIGatewayProxyEvent
): Promise<ParsedRequest> {
    const language = event.pathParameters?.lang;
    if (!isSupportedLanguage(language)) {
        throw new Error('Unsupported language');
    }
    const body = qs.parse(event.body || '');
    const Caller = body.Caller || event.queryStringParameters?.Caller;
    if (typeof Caller !== 'string' || Caller.length < 3) {
        throw new Error('Unable to identify caller');
    }
    const user = await getItem(Caller);
    return {
        language,
        user,
        event,
    };
}

export const safeHandle = (
    f: (e: ParsedRequest) => Promise<VoiceResponse>
): APIGatewayProxyHandler => {
    const safeHandler = async function safelyHandledFunction(
        e: APIGatewayProxyEvent
    ) {
        try {
            const result = await parseRequest(e);
            const response = await f(result);
            return {
                statusCode: 200,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: response.toString(),
            };
        } catch (err) {
            const maybeLang = e.pathParameters?.lang;
            const language = isSupportedLanguage(maybeLang)
                ? maybeLang
                : 'en-GB';
            const response = new twiml.VoiceResponse();
            const message =
                typeof err === 'string'
                    ? err
                    : typeof err.message === 'string'
                    ? err.message
                    : '';
            response.say(
                getVoiceParams(language),
                __('error', { message }, language)
            );
            return {
                statusCode: 500,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: response.toString(),
            };
        }
    };
    safeHandler.orig = f;
    return safeHandler;
};
