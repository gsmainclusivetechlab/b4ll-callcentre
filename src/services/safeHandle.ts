import {
    APIGatewayProxyHandler,
    APIGatewayProxyEvent,
    APIGatewayProxyResult,
} from 'aws-lambda';
import {
    getVoiceParams,
    __,
    isSupportedLanguage,
    SupportedLanguage,
} from './strings';
import { twiml } from 'twilio';
import { getAccountItem, RecordType } from './dynamodb';
import qs from 'querystring';
import * as jwt from 'jsonwebtoken';
import {
    HandlerParams,
    handleVerification,
    AuthCookie,
    VerificationState,
} from './auth';

export interface ParsedRequest {
    language: SupportedLanguage;
    user: RecordType;
    event: APIGatewayProxyEvent;
    auth: AuthCookie;
}

async function parseRequest(
    event: APIGatewayProxyEvent
): Promise<ParsedRequest> {
    console.log('parsing', event);
    const language = event.pathParameters?.lang;
    if (!isSupportedLanguage(language)) {
        throw new Error('Unsupported language');
    }
    const body = qs.parse(event.body || '');
    const callerKey =
        body.Direction === 'outbound-api' ||
        event.queryStringParameters?.Direction === 'outbount-api'
            ? 'Called'
            : 'Caller';
    const Caller = body[callerKey] || event.queryStringParameters?.[callerKey];
    if (typeof Caller !== 'string' || Caller.length < 3) {
        throw new Error('Unable to identify caller');
    }
    const user = await getAccountItem(Caller);

    // By default, the user is unauthenticated
    let auth: AuthCookie = {
        sub: Caller,
        state: user.isEnrolled
            ? VerificationState.REGISTERED
            : VerificationState.NOT_ENROLLED,
    };
    const cookie = event.headers.Cookie?.split(/;\s?/)
        ?.map((x) => x.split('='))
        ?.find(([name]) => name === 'Auth')?.[1];
    if (cookie) {
        try {
            if (!process.env.JWT_KEY)
                throw new Error('No JWT Key found in environment');
            auth = jwt.verify(cookie, process.env.JWT_KEY, {
                subject: Caller,
            }) as AuthCookie;
        } catch (e) {
            console.log(e);
            // ignore the failure and treat this as an unauthorized user
        }
    }
    console.log('parse result ', { language, user, event, auth });

    // normalise event path
    if (event.requestContext) {
        event.path = event.requestContext.path;
        (event as any).domainName = event.requestContext.domainName;
    }

    return {
        language,
        user,
        event,
        auth,
    };
}

export const safeHandle = (
    f: (e: ParsedRequest) => Promise<{ toString: () => string }>,
    params: HandlerParams = {}
): APIGatewayProxyHandler => {
    const safeHandler = async function safelyHandledFunction(
        e: APIGatewayProxyEvent
    ) {
        try {
            const request = await parseRequest(e);
            if (params.requireVerification) {
                const result = await handleVerification(request, params);
                if (result) return result;
            }

            const response = await f(request);
            return serialise(response, 200);
        } catch (err) {
            const maybeLang = e.pathParameters?.lang;
            const language = isSupportedLanguage(maybeLang)
                ? maybeLang
                : 'en-GB';
            if (typeof err === 'object') {
                return serialise(err, err.statusCode || 500);
            }
            const response = new twiml.VoiceResponse();
            const message =
                typeof err === 'string'
                    ? err
                    : typeof err.message === 'string'
                    ? err.message
                    : '';
            response.say(
                getVoiceParams(language),
                __('error', { message }, language)
            );
            return serialise(response, 500);
        }
    };
    safeHandler.orig = f;
    return safeHandler;
};

function serialise(r: unknown, statusCode: number): APIGatewayProxyResult {
    console.log('Returning ', r);
    if (typeof r === 'object' && r !== null) {
        if (r instanceof twiml.VoiceResponse) {
            return {
                statusCode,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: r.toString(),
            };
        }
    }
    return {
        statusCode,
        headers: {
            'Content-Type': 'text/json',
            'access-control-allow-origin': '*',
        },
        body: JSON.stringify(r),
    };
}
