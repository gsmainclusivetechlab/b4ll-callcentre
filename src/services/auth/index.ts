import { APIGatewayProxyResult } from 'aws-lambda';
import { getVoiceParams, __ } from '../strings';
import { twiml } from 'twilio';
import { requestVerification } from './requestVerification';
import { checkVerification } from './checkVerification';
import { handleEnrolment } from './handleEnrolment';
import { requestEnrolment } from './requestEnrolment';
import { ParsedRequest } from '../safeHandle';
import * as jwt from 'jsonwebtoken';
import {
    VoiceItEnrolmentData,
    VoiceItVerificationData,
} from '../../engine/voiceit/provider';

export enum VerificationState {
    NOT_ENROLLED = 0,
    ENROLMENT_REQUESTED = 1,
    REGISTERED = 2,
    AUTHENTICATION_REQUESTED = 3,
    AUTHENTICATED = 4,
    DEACTIVATED = 5,
}
interface BaseCookie {
    /** caller phone number */
    sub: string;
    exp?: number;
}
interface PlainCookie extends BaseCookie {
    state:
        | VerificationState.AUTHENTICATED
        | VerificationState.NOT_ENROLLED
        | VerificationState.REGISTERED
        | VerificationState.DEACTIVATED;
}
interface EnrolmentRequestedCookie extends BaseCookie {
    state: VerificationState.ENROLMENT_REQUESTED;
    /** phrase that the user was asked to record */
    phrase: string;
    req: VoiceItEnrolmentData;
}
interface AuthenticationRequestedCookie extends BaseCookie {
    state: VerificationState.AUTHENTICATION_REQUESTED;
    req: VoiceItVerificationData;
}

export type AuthCookie =
    | PlainCookie
    | AuthenticationRequestedCookie
    | EnrolmentRequestedCookie;

export function makeCookieHeader(cookie: AuthCookie): Record<string, string> {
    const signed = jwt.sign(cookie, process.env.JWT_KEY || '', {
        expiresIn: '30s',
    });
    return {
        'Set-Cookie': `Auth=${signed}`,
    };
}

export interface HandlerParams {
    requireVerification?: boolean;
    allowEnrolment?: boolean;
    allowDeactivated?: boolean;
    loginRedirect?: { method: string; target: string };
}

export async function handleVerification(
    request: ParsedRequest,
    params: HandlerParams
): Promise<null | APIGatewayProxyResult> {
    const { auth, language } = request;
    console.log('Verifying, user is in state ', auth);
    switch (auth?.state) {
        default:
        case VerificationState.NOT_ENROLLED:
            if (params.allowEnrolment === false) break;
            return requestEnrolment(request);
        case VerificationState.ENROLMENT_REQUESTED:
            return handleEnrolment(request, params.loginRedirect);
        case VerificationState.REGISTERED:
            return requestVerification(request);
        case VerificationState.AUTHENTICATION_REQUESTED:
            return checkVerification(request, params.loginRedirect);
        case VerificationState.AUTHENTICATED:
            return null;
        case VerificationState.DEACTIVATED:
            if (params.allowDeactivated === true) return null; // proceed to the original function handler
            break; // proceed to the unauthorised handler below
    }

    // currently the only way to arrive here is if enrolment is required but not allowed on the handler
    const response = new twiml.VoiceResponse();
    response.say(getVoiceParams(language), __('unauthorised', language));
    return {
        statusCode: 200,
        headers: {
            'Content-Type': 'text/xml',
        },
        body: response.toString(),
    };
}
