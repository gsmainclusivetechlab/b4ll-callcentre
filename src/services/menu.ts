import { twiml } from 'twilio';
import { getVoiceParams, __ } from './strings';
import { ParsedRequest } from './safeHandle';
import querystring from 'querystring';
import { MessageId } from '../strings';
import VoiceResponse from 'twilio/lib/twiml/VoiceResponse';

export interface MenuOption {
    triggers: string[];
    description: MessageId | ((r: ParsedRequest) => Promise<MessageId>);
    handler: (r: ParsedRequest) => Promise<VoiceResponse>;
}

export async function menuToHandler(
    menu: MenuOption[],
    { event, language, user, auth }: ParsedRequest,
    redirect?: string
): Promise<VoiceResponse> {
    const { Digits, SpeechResult } = querystring.parse(event.body || '');
    const answer = Digits || SpeechResult || null;
    if (typeof answer === 'string') {
        const selected = menu.find(
            (option, i) =>
                answer === '' + (i + 1) || option.triggers.indexOf(answer) >= 0
        );
        if (selected) {
            return selected.handler({ language, user, event, auth });
        }
    }

    // redirect to root and we can try again
    const response = new twiml.VoiceResponse();
    response.say(getVoiceParams(language), __('did-not-understand', language));
    if (redirect) response.redirect({ method: 'GET' }, redirect);
    return response;
}

export async function menuToGather(
    response: VoiceResponse,
    request: ParsedRequest,
    menu: MenuOption[]
): Promise<VoiceResponse.Gather> {
    const voice = getVoiceParams(request.language);
    const gather = response.gather({
        input: ['dtmf'],
        method: 'POST',
        language: voice.language,
        hints: menu.flatMap((o) => o.triggers).join(','),
        numDigits: 1,
        timeout: 5,
    });

    // TODO: consider better behaviour in empty menu case and > 9 items case
    for (let i = 0; i < Math.min(menu.length, 9); i++) {
        const { description: maybeDesc } = menu[i];
        const description: MessageId =
            typeof maybeDesc === 'string'
                ? maybeDesc
                : await maybeDesc(request);
        gather.say(voice, __(description, { index: i + 1 }, request.language));
    }

    return gather;
}
