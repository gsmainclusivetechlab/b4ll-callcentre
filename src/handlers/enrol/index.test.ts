import * as handler from '.';
import { mockHandlerFn } from '../../../dev/mockHandlerFn';
import qs from 'querystring';
import * as voiceit from '../../services/voiceit';
import * as dynamodb from '../../services/dynamodb';
const { isLegalPhraseKey } = jest.requireActual('../../services/voiceit');

jest.mock('../../services/voiceit');
jest.mock('../../services/dynamodb');

const createUser = voiceit.createUser as jest.MockedFunction<
    typeof voiceit.createUser
>;
const enrolUser = voiceit.enrolUser as jest.MockedFunction<
    typeof voiceit.enrolUser
>;
const putItem = dynamodb.putItem as jest.MockedFunction<
    typeof dynamodb.putItem
>;

(voiceit.isLegalPhraseKey as jest.MockedFunction<
    typeof voiceit.isLegalPhraseKey
>).mockImplementation(isLegalPhraseKey);

const get = mockHandlerFn(handler.get);
const post = mockHandlerFn(handler.post);

const mockResponse = {
    message: 'Successfully achieved objective',
    status: 200,
    timeTaken: '0.0',
    responseCode: 'SUCC' as const,
    contentLanguage: 'en-DEV',
    createdAt: 0,
    text: 'zoo-phrase',
    voiceEnrolmentId: '0',
};

describe('enrolment flow', () => {
    beforeEach(() => {
        createUser.mockReset();
        enrolUser.mockReset();
        putItem.mockReset();
    });

    test('should create a user when necessary', async () => {
        createUser.mockResolvedValue({
            ...mockResponse,
            userId: 'id_voiceit',
            createdAt: 0,
        });
        let result = await get({
            language: 'en-DEV',
            user: { id: '+77-enrol-get-test' },
        });
        expect(createUser).toHaveBeenCalled();
        expect(putItem).toHaveBeenCalledWith({
            id: '+77-enrol-get-test',
            voiceItId: 'id_voiceit',
        });
        createUser.mockReset();
        putItem.mockReset();
        result = await get({
            language: 'en-DEV',
            user: { id: '+77-enrol-get-test', voiceItId: 'id_voiceit' },
        });
        expect(createUser).not.toHaveBeenCalled();
        expect(putItem).not.toHaveBeenCalled();

        expect(result.toString()).toContain('enrol-message');
        expect(result.toString()).toContain('tomorrow-phrase');
        expect(result.toString()).toContain('./enrol/Tomorrow');
        expect(result.toString()).toMatchSnapshot();
    });
    test.each`
        phraseKey | hasUser  | hasRecording | expected
        ${null}   | ${true}  | ${true}      | ${'Unrecognised phrase key'}
        ${'Zoo'}  | ${true}  | ${false}     | ${'Could not retrieve recording URL.'}
        ${'Zoo'}  | ${false} | ${true}      | ${'VoiceIT user does not exist'}
        ${'Zoo'}  | ${true}  | ${true}      | ${'Enrolment failed'}
    `(
        'should handle error "$expected"',
        async ({ phraseKey, hasUser, hasRecording, expected }) => {
            enrolUser.mockResolvedValue({
                ...mockResponse,
                responseCode: 'FAIL',
            });
            const response = post({
                language: 'en-DEV',
                user: {
                    id: '+77-post-test',
                    ...(hasUser ? { voiceItId: '123' } : null),
                },
                event: {
                    ...(phraseKey ? { pathParameters: { phraseKey } } : null),
                    body: qs.stringify(
                        hasRecording ? { RecordingUrl: 'file.wav' } : {}
                    ),
                },
            });
            await expect(response).rejects.toThrow(expected);
        }
    );

    // test.each`
    //     phraseKey | user                    | body                           | result | expected
    //     ${'fake'} | ${{}}                   | ${{}}                          | ${{}}  | ${new Error('Unrecognised phrase key')}
    //     ${'Zoo'}  | ${{}}                   | ${{}}                          | ${{}}  | ${new Error('Could not retrieve recording URL.')}
    //     ${'Zoo'}  | ${{}}                   | ${{ RecordingUrl: 'foo.wav' }} | ${{}}  | ${new Error('VoiceIT user does not exist')}
    //     ${'Zoo'}  | ${{ voiceItId: '123' }} | ${{ RecordingUrl: 'foo.wav' }} | ${{}}  | ${new Error('Enrolment Failed')}
    // `(
    test('should collect first enrolments', async () => {
        enrolUser.mockResolvedValue(mockResponse);

        const response = await post({
            language: 'en-DEV',
            user: { id: '+77-enrol-post-test', voiceItId: 'v' },
            event: {
                pathParameters: { phraseKey: 'Zoo' },
                body: qs.stringify({ RecordingUrl: 'file.wav' }),
            },
        });

        expect(putItem).toHaveBeenCalledWith({
            id: '+77-enrol-post-test',
            voiceItId: 'v',
            enrolments: { Zoo: 1 },
            isEnrolled: false,
        });

        expect(response.toString()).toContain(
            'enrol-confirmation: 2 remaining'
        );
        expect(response.toString()).toContain('zoo-phrase');
    });

    test('should collect last enrolments', async () => {
        enrolUser.mockResolvedValue(mockResponse);

        const response = await post({
            language: 'en-DEV',
            user: {
                id: '+77-enrol-post-test',
                voiceItId: 'v',
                enrolments: { Zoo: 2 },
            },
            event: {
                pathParameters: { phraseKey: 'Zoo' },
                body: qs.stringify({ RecordingUrl: 'file.wav' }),
            },
        });

        expect(putItem).toHaveBeenCalledWith({
            id: '+77-enrol-post-test',
            voiceItId: 'v',
            enrolments: { Zoo: 3 },
            isEnrolled: true,
        });

        expect(response.toString()).toContain('enrolment-complete');
        expect(response.toString()).toContain('../menu');
    });
});
