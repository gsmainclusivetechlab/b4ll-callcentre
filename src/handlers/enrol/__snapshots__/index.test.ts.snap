// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enrolment flow should create a user when necessary 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Emma-Neural">
        enrol-message
    </Say>
    <Pause length="1"/>
    <Say language="en-GB" voice="Polly.Emma-Neural">
        tomorrow-phrase
    </Say>
    <Record action="./enrol/Tomorrow" method="POST" finishOnKey="#" playBeep="true" maxLength="5"/>
</Response>
`;
